### PR TITLE
[bug] Update TVirtualX.cxx to fix 6483

### LINKED ISSUE #6483: [GUI] Crashes when creating TGClient in a standalone application
When creating a TGClient instance in a standalone application. ROOT crashes. 
This reproduces the problem:
```
#include "TGClient.h"
#include "TROOT.h"
#include <iostream>

void test() {
   std::cout << " ROOT is in batch " << gROOT->IsBatch()  << std::endl;
   auto g = new TGClient(); 
}

int main() {
   test(); 
} 

This makes the fix applied for https://sft.its.cern.ch/jira/browse/ROOT-9483 not working anymore in the current master

### PR DESCRIPTION
**Pull request by andriish**
New: This should fix #6483

Old:
This will partially fix #6483
The full fix should include 
```
  if (fDefaultFont) gval.fFont       = fDefaultFont->GetFontHandle();
...
  if(fDocPropFont) gval.fFont       = fDocPropFont->GetFontHand
...   
  if (fStatusFont) gval.fFont       = fStatusFont->GetFontHandle();
```
in the TGResourcePool.cxx
However, I don't know if that makes sense, so I don't include it.

Tag @couet  @lmoneta

### CODE REVIEW

**Comment by andriish** [comment]
I mean of course 

```
if (fStatusFont) gval.fFont       = fStatusFont->GetFontHandle();
```
should be there, but I don't know if there should be something like
```
if (fStatusFont) gval.fFont       = fStatusFont->GetFontHandle(); else   gval.fFont       = ... 
```

**Comment by couet** [comment]
You said:
This will partially fix #6543
But #6543 is a closed PR .... not an issue ...

**Comment by andriish** [comment]
Hi @couet 

Yes, I mean:

#6483 

Clicked at wrong  github popping suggestion.


Best regards,
Andrii

File: gui/gui/src/TGResourcePool.cxx [hunk 1]
```diff
@@ -168,7 +168,7 @@ TGResourcePool::TGResourcePool(TGClient *client)
                 kGCFillStyle  | kGCGraphicsExposures;
    gval.fGraphicsExposures = kFALSE;
    gval.fFillStyle  = kFillSolid;
-   gval.fFont       = fDefaultFont->GetFontHandle();
+   gval.fFont       = fDefaultFont?fDefaultFont->GetFontHandle():FontH_t(0);
```
Comment (bellenot) [inline gui/gui/src/TGResourcePool.cxx:171]:
```suggestion
   gval.fFont       = fDefaultFont ? fDefaultFont->GetFontHandle() : FontH_t(0);
```

File: gui/gui/src/TGResourcePool.cxx [hunk 2]
```diff
@@ -195,15 +195,15 @@ TGResourcePool::TGResourcePool(TGClient *client)
    gval.fBackground = fSelBackColor;
    fSelGC = fGCPool->GetGC(&gval, kTRUE);
 
-   gval.fFont       = fDocPropFont->GetFontHandle();
+   gval.fFont       = fDocPropFont?fDocPropFont->GetFontHandle():FontH_t(0);
```
Comment (bellenot) [inline gui/gui/src/TGResourcePool.cxx:198]:
```suggestion
   gval.fFont       = fDocPropFont ? fDocPropFont->GetFontHandle() : FontH_t(0);
```

File: gui/gui/src/TGResourcePool.cxx [hunk 3]
```diff
@@ -195,15 +195,15 @@ TGResourcePool::TGResourcePool(TGClient *client)
...
    gval.fForeground = fDocForeColor;
    gval.fBackground = fDocBackColor;
    fDocGC = fGCPool->GetGC(&gval, kTRUE);
 
    gval.fForeground = fDocBackColor;
    fDocbgndGC = fGCPool->GetGC(&gval, kTRUE);
 
-   gval.fFont       = fStatusFont->GetFontHandle();
+   gval.fFont       = fStatusFont?fStatusFont->GetFontHandle():FontH_t(0);
```
Comment (bellenot) [inline gui/gui/src/TGResourcePool.cxx:206]:
```suggestion
   gval.fFont       = fStatusFont ? fStatusFont->GetFontHandle() : FontH_t(0);
```

**Review by bellenot — CHANGES_REQUESTED** [review]
Please give more details on what your patch is fixing

**Comment by andriish** [comment]
Hi @bellenot ,

The patch fixes the crash in reproducer in #6483 .

The reason of the crash is that in a batch mode the ListFonts() from TVirtualX is called. The current version of ListFonts() does not return 0 fonts (but it should!)  and then the functions that try to deal with fonts crash everything.
Right now I cannot recall the exact call chain that leads to crash, but here are just some cases of undefined behaviour:
```
char **TGFontPool::GetFontFamilies()
{
   Int_t i, numNames;
   char *family, *end, *p;

   THashTable familyTable(100);
   familyTable.SetOwner();

   char **nameList;
   char **dst;

   // coverity[returned_null]
   // coverity[dereference]
   nameList = gVirtualX->ListFonts("*", 10000, numNames);

   for (i = 0; i < numNames; i++) {       <-- numNames is not initialized explicitly 

```

The second part of the fix is in the TGResourcePool.cxx . Basically the fix is for the case  when no fonts were given back, i.e. just avoiding null pointer dereference.

Best regards,

Andrii

**Comment by bellenot** [comment]
OK, I see, thanks for the clarification @andriish !

**Review by bellenot — APPROVED** [review]

**Pull request merged**